src/main/java/io/jenkins/plugins/artifact_manager_s3/S3BlobStore.java

```diff
@@ -75,25 +76,49 @@
 
     private static final long serialVersionUID = -8864075675579867370L;
 
-    // For now, these are taken from the environment, rather than being configured.
-    @SuppressWarnings("FieldMayBeFinal")
-    private static String BLOB_CONTAINER = System.getenv("S3_BUCKET");
-    @SuppressWarnings("FieldMayBeFinal")
-    private static String PREFIX = System.getenv("S3_DIR");
-    @SuppressWarnings("FieldMayBeFinal")
-    private static String REGION = System.getProperty(S3BlobStore.class.getName() + ".region");
+    public static final String KEY_CONTAINER = S3BlobStore.class.getName() + ".container";
+    public static final String KEY_PREFIX = S3BlobStore.class.getName() + ".prefix";
+    public static final String KEY_REGION = S3BlobStore.class.getName() + ".region";
+
+    public final String PREFIX_PROPERTY = System.getProperty(KEY_PREFIX);
+    public final String CONTAINER_PROPERTY = System.getProperty(KEY_CONTAINER);
+    public final String REGION_PROPERTY = System.getProperty(KEY_REGION);
+
+    private String container;
+    private String prefix;
+    private String region;
 
     @DataBoundConstructor
-    public S3BlobStore() {}
+    public S3BlobStore() {
+    }
 
     @Override
     public String getPrefix() {
-        return PREFIX;
+        return StringUtils.defaultIfBlank(prefix, PREFIX_PROPERTY);
     }
 
     @Override
     public String getContainer() {
-        return BLOB_CONTAINER;
+        return StringUtils.defaultIfBlank(container, CONTAINER_PROPERTY);
```

Comment:
I do not think we should do it this way at all.

Just have the user configure all required fields (I suppose container, maybe prefix) and pass them in the `@DataBoundConstructor` treating them as mandatory (throw an exception if null/blank). Any optional fields (region, and perhaps prefix) can be in `@DataBoundSetter`s. (The empty string would be a reasonably default value for the prefix.)

System properties may be used to preconfigure _test_ code. That means reading them in `S3AbstractTest`.

Reply:
is this compatible with JCasC? this change the previous behaviour where you defined environment variables as configuration input